core/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardAllocationDecision.java

```diff
@@ -72,47 +90,53 @@ private UnassignedShardDecision(Decision finalDecision,
+        if (explanation != null) {
+            return new ShardAllocationDecision(Type.NO, allocationStatus, explanation, null, null, nodeDecisions);
+        } else {
+            return getCachedDecision(allocationStatus);
+        }
     }
 
     /**
-     * Creates a THROTTLE decision with the given explanation and individual node-level decisions that
-     * comprised the final THROTTLE decision.
+     * Returns a THROTTLE decision, with the given explanation and individual node-level decisions that
+     * comprised the final THROTTLE decision if in explain mode.
      */
-    public static UnassignedShardDecision throttleDecision(String explanation,
-                                                           Map<String, Decision> nodeDecisions) {
-        Objects.requireNonNull(explanation, "explanation must not be null");
-        return new UnassignedShardDecision(Decision.THROTTLE, AllocationStatus.DECIDERS_THROTTLED, explanation, null, null,
-                                           nodeDecisions);
+    public static ShardAllocationDecision throttle(@Nullable String explanation, @Nullable Map<String, Decision> nodeDecisions) {
+        if (explanation != null) {
+            return new ShardAllocationDecision(Type.THROTTLE, AllocationStatus.DECIDERS_THROTTLED, explanation, null, null, nodeDecisions);
+        } else {
+            return getCachedDecision(AllocationStatus.DECIDERS_THROTTLED);
+        }
     }
 
     /**
      * Creates a YES decision with the given explanation and individual node-level decisions that
      * comprised the final YES decision, along with the node id to which the shard is assigned and
      * the allocation id for the shard, if available.
      */
-    public static UnassignedShardDecision yesDecision(String explanation,
-                                                      String assignedNodeId,
-                                                      @Nullable String allocationId,
-                                                      Map<String, Decision> nodeDecisions) {
-        Objects.requireNonNull(explanation, "explanation must not be null");
+    public static ShardAllocationDecision yes(String assignedNodeId, @Nullable String explanation, @Nullable String allocationId,
+                                              @Nullable Map<String, Decision> nodeDecisions) {
         Objects.requireNonNull(assignedNodeId, "assignedNodeId must not be null");
-        return new UnassignedShardDecision(Decision.YES, null, explanation, assignedNodeId, allocationId, nodeDecisions);
+        return new ShardAllocationDecision(Type.YES, null, explanation, assignedNodeId, allocationId, nodeDecisions);
+    }
+
+    private static ShardAllocationDecision getCachedDecision(AllocationStatus allocationStatus) {
+        ShardAllocationDecision decision = CACHED_DECISIONS.get(allocationStatus);
+        assert decision != null : "precomputed decision not found for " + allocationStatus;
```

Comment:
I'm on the fence about making this a true `Objects.requireNonNull(decision, "...")`, what do you think?